### PR TITLE
Fix error handling for Tap CLI and public<>tap server

### LINKED ISSUE #106: Extract and test protobuf-over-http code
As part of #103, our client/server implementations for the public API were refactored, but they're still pretty convoluted. Specifically, the protobuf-over-http setup is brittle and not thoroughly tested. We should move that code to a separate package and add additional tests.

### PR DESCRIPTION
**Pull request by pcalcado**
-  first step towards #49
- closes #106

This started as an investigation of why `$ conduit tap` would abort with `unexpected EOF` at times.

To debug, I've read all the protobuf-to-http code and, while at that I've cleaned up it a bit and wrote tests for a lot of it.

I finally identified that this problem was due to us not checking when the gRPC `Tap()` returned an error (in this case because the pods didnt exist). Commit c27f400e8918d94ce566f63dea94d4af74c2ab7a  adds  this check.

The EOF goes away in the client, but the error message still wasn't great:
```
pcalcado@arbeitsziege.local:~/code/src/github.com/runconduit/conduit  $ conduit tap pod default/pod-with-no-rules  --api-addr localhost:8085
rpc error: code = Unknown desc = no pod exists for key default/pod-with-no-rules%
```

So I did some digging and realized that we haven't been using gRPC errors properly. Commit a3954e6 adds that to the tap server, and after it here is the message displayed:

```
pcalcado@arbeitsziege.local:~/code/src/github.com/runconduit/conduit  $ conduit tap pod default/pod-with-no-rules  --api-addr localhost:8085
no pod exists for key default/pod-with-no-rules%
```

### CODE REVIEW

**Comment by pcalcado** [comment]
FWIW, tests arent passing on travis and I am debugging it so this isnt super ready for review yet :(

**Comment by pcalcado** [comment]
Alright, adding `review` & `reviewable` to this, it should be good to go!

File: controller/api/public/client.go [hunk 1]
```diff
@@ -109,11 +89,13 @@ func (c *grpcOverHttpClient) apiRequest(ctx context.Context, endpoint string, re
 		return fmt.Errorf("POST to Conduit API endpoint [%s] returned HTTP status [%s]", url, httpRsp.Status)
 	}
 
-	defer httpRsp.Body.Close()
+	if err = checkIfResponseHasConduitError(httpRsp); err != nil {
+		return err
+	}
 
+	defer httpRsp.Body.Close()
```
[the remark this answers is not in the document]
Comment (pcalcado) [reply controller/api/public/client.go]: I'm not sure. there's a lot of error handling before we get the `Body`, I would assume that we should "group" the lines that handle that object. Maybe this is a code smell indicating that this should be extracted in a func?

**Comment by pcalcado** [comment]
@siggy you are right, sorry about the PR size. 

I think that I need to build better intuition for this. Both coming from a trunk-based development background and being a little worried about review turnaround time bias me towards larger and more end-to-end pieces, but I'll change that 💪

File: controller/api/public/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (siggy) [reply controller/api/public/client.go]: my concern was that if we return prior to this, we'll resource leak. reading this code a bit more i'm realizing `checkIfResponseHasConduitError` also calls close. maybe the safest thing to do is to `httpRsp.Body.Close()` immediately after the `c.post(...` error check, and then `checkIfResponseHasConduitError()` need not be concerned with closing?

**Comment by pcalcado** [comment]
@klingerf do you still want to review this?

**Comment by klingerf** [comment]
@pcalcado thanks for checking -- I should be able to get to it shortly.

File: controller/api/public/http_server.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	counter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "http_requests_total",
+			Help: "A counter for requests to the wrapped handler.",
+		},
+		[]string{"code", "method"},
+	)
+	prometheus.MustRegister(counter)
+	instrumentedHandler := promhttp.InstrumentHandlerCounter(counter, baseHandler)
+	return instrumentedHandler
```
Comment (klingerf) [inline controller/api/public/http_server.go]: TIOLI No need for the `instrumentedHandler` local variable here -- you can just return the result of `InstrumentHandlerCounter` directly:

```
return promhttp.InstrumentHandlerCounter(counter, baseHandler)
```

File: controller/api/public/http_server_test.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+	httpServer := &http.Server{
+		Addr:    "localhost:8889",
+		Handler: handler,
+	}
+
+	go func() {
+		err := httpServer.ListenAndServe()
+		if err != nil {
+			fmt.Println(err)
```
Comment (klingerf) [inline controller/api/public/http_server_test.go]: I don't think this will show up in standard `go test` output. I recommend using `t.Log(err)` instead (or, alternatively, removing in since the error is logged in the next line anyway).

File: controller/api/public/http_server_test.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+			t.Fatalf("Could not start server: %v", err)
+		}
+	}()
+
+	client, err := NewInternalClient("localhost:8889")
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	t.Run("Delegates all non-streaming RPC messages to the underlying grpc server", func(t *testing.T) {
```
Comment (klingerf) [inline controller/api/public/http_server_test.go]: TIOLI I'm inclined to split this into 4 separate tests -- one for each of the non-streaming endpoints.

File: controller/api/public/proto_over_http.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+import (
+	"bufio"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"google.golang.org/grpc/status"
+
```
Comment (klingerf) [inline controller/api/public/proto_over_http.go]: Extra newline

File: controller/api/public/proto_over_http_test.go [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+
+func newStubResponseWriter() *stubResponseWriter {
+	return &stubResponseWriter{
+		headers: make(http.Header),
+		body:    bytes.NewBufferString(""),
+	}
+}
+
+func TestHttpRequestToProto(t *testing.T) {
+	someUrl := "www.example.org/something"
```
Comment (klingerf) [inline controller/api/public/proto_over_http_test.go]: TIOLI maybe use a fully-qualified URL here?

File: controller/api/public/proto_over_http_test.go [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+
+		expectedErrorPayload := pb.ApiError{Error: genericError.Error()}
+		var actualErrorPayload pb.ApiError
+		err = proto.Unmarshal(payloadRead, &actualErrorPayload)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		if actualErrorPayload != expectedErrorPayload {
+			t.Fatalf("Expecting error to be serialised as [%v], but got [%v]", expectedErrorPayload, actualErrorPayload)
```
Comment (klingerf) [inline controller/api/public/proto_over_http_test.go]: Typo here: `serialised` => `serialized`. There are a few other occurrences in this file as well.

File: controller/api/public/proto_over_http_test.go [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+		}
+
+		if expectedMessage != actualMessage {
+			t.Fatalf("Expected response body to contain message [%v], but got [%v]", expectedMessage, actualMessage)
+		}
+	})
+}
+
+func TestPayloadSize(t *testing.T) {
+	t.Run("Can write and read message correctly based on payload size correct payload size to message", func(t *testing.T) {
```
Comment (klingerf) [inline controller/api/public/proto_over_http_test.go]: This test description is a bit hard to parse. Maybe it should be:

> Can write and read message correctly based on payload size

File: controller/api/public/proto_over_http_test.go [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+	t.Run("Returns an error if writer doesnt support streaming", func(t *testing.T) {
+		_, err := newStreamingWriter(&nonStreamingResponseWriter{})
+		if err == nil {
+			t.Fatalf("Expecting error, got nothing")
+		}
+	})
+}
+
+func TestCheckIfResponseHasError(t *testing.T) {
+	t.Run("If response doesnt contain Conduit error, returns nil", func(t *testing.T) {
```
Comment (klingerf) [inline controller/api/public/proto_over_http_test.go]: TIOLI I'd reverse the three test descriptions inside this function to put the conditional at the end, since I think that makes them more readable.

> Returns nil if response doesn't contain Conduit error

> Returns error from body if response contains Conduit error

> Returns error if response contains Conduit error but body isn't error message

File: controller/api/public/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pcalcado) [reply controller/api/public/client.go]: I think I know what makes me feel funny about this: the `post` call "owns" the `Response`, but not the `Body`. The `Body` is only used in thos few lines at the bottom. I think that a good tool to think about this is that if we were to extract those few lines into a `convertBodyIntoProto` or something the `defer` would be extracted as well.

Nevertheless I thing the suggestions are good enough and just went ahead with them 👍

File: controller/api/public/client_test.go [hunk 10]
```diff
@@ -70,37 +70,37 @@ func TestFromByteStreamToProtocolBuffers(t *testing.T) {
...
-	t.Run("Correctly marshalls a large byte arrey", func(t *testing.T) {
+	t.Run("Correctly marshalls a large byte array", func(t *testing.T) {
 		series := pb.MetricSeries{
 			Name:       pb.MetricName_REQUEST_RATE,
 			Metadata:   &pb.MetricMetadata{},
 			Datapoints: make([]*pb.MetricDatapoint, 0),
 		}
 
-		numberOfDatapointsInMessage := 1000
+		numberOfDatapointsInMessage := 50
```
[the remark this answers is not in the document]
Comment (pcalcado) [reply controller/api/public/client_test.go]: Aha! Now I understand it. Let me enconde this knowledge in the code.

File: controller/api/public/http_server_test.go (hunk 3 above)
Comment (pcalcado) [reply controller/api/public/http_server_test.go]: It will with `-v` but tbh I'm pretty sure this is vestigial from me debugging some problems I had running these tests on travis so I'll remove it.

File: controller/api/public/proto_over_http.go [hunk 11]
```diff
@@ -118,32 +118,18 @@ func serializeAsPayload(messageContentsInBytes []byte) ([]byte, error) {
...
-			break
-		}
+	messageContentsAsBytes := make([]byte, messageLength)
+	totalBytesRead, err := io.ReadFull(reader, messageContentsAsBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	if totalBytesRead != messageLength {
+		return nil, fmt.Errorf("message declared length [%d] but could only read [%d] bytes")
```
Comment (deebo91) [inline controller/api/public/proto_over_http.go]: Just to wrap my head around this error case, What scenarios would cause the code to reach this line when deserializing the HTTP? Proto? body.

**Comment by pcalcado** [comment]
@deebo91 [said](https://github.com/runconduit/conduit/pull/177#discussion_r163688318):
```go
if totalBytesRead != messageLength {
 return nil, fmt.Errorf("message declared length [%d] but could only read [%d] bytes")
}
```
> Just to wrap my head around this error case, What scenarios would cause the code to reach this line when deserializing the HTTP? Proto? body.


The protocol we use here is custom, and it relies on declaring the message length using [the first four bytes of the HTTP body](https://github.com/runconduit/conduit/blob/5e6684f9f28abe46160ad2c5b8049bf9cb5b9a29/controller/api/public/proto_over_http.go#L114). If everything goes well, the server code probably uses `serializeAsPayload` to encode the message and everybody will be happy.

Unfortunately the world is a sad place. It could be that:
-  we accidentally introduce a code path on the server that doesn't call `serializeAsPayload` but rather use a copy adn paste'd version that isn't broken
- we might want to change the protocol in an upcoming version (say using more or fewer bytes than 4) and we might have a situation where the user has a CLI version that is a bit older and doesn't know about the protocol change
- When writing distributed systems, even if you have absolute control of the `from` and `to` components you still have to deal with the unreliable network. There are many pieces of middleware between the CLI and the actual server, from the k8s API to k8s itself to maybe even a web accelerator proxy thing or some corporate spyware that the user might have installed in their network. These things, especially some weird proprietary "security" tools corporations sometimes use, can do all sort of transformations with they payload. It might be the case that you hit a defective one and they strip out some bytes from your message. If we don't check this here, we will have a very confusing error when this gets unmarshalled. Instead, we should detect the error as soon as possible and inform this function's caller that something isn't cool.

File: controller/api/public/proto_over_http_test.go (hunk 7 above)
Comment (pcalcado) [reply controller/api/public/proto_over_http_test.go]: 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺 🇬🇧 🇦🇺

File: controller/api/public/proto_over_http_test.go [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		if string(actualMessage) != expectedMessage {
+			t.Fatalf("Expecting payload to contain message [%s], but it had [%s]", expectedMessage, actualMessage)
+		}
+	})
+
+	t.Run("Can multiple messages in the same stream", func(t *testing.T) {
```
Comment (klingerf) [inline controller/api/public/proto_over_http_test.go]: Hmm, maybe:

> Can _read_ multiple messages in the same stream
